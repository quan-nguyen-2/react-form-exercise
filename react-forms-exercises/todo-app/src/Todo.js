import React, { useState } from "react";

const Todo = ({ id, task, update, remove }) => {
  const [isEditing, setIsEditing] = useState(false);
  const [editTask, setEditTask] = useState(task);
  const [completed, setCompleted] = useState(false);

  const toggleCompletion = () => {
    completed ? setCompleted(false) : setCompleted(true);
  };

  const toggleEdit = () => {
    setIsEditing((edit) => !edit);
  };

  const handleChange = (e) => {
    setEditTask(e.target.value);
  };

  const handleUpdate = (e) => {
    e.preventDefault();
    update(id, editTask);
    setIsEditing(false);
  };

  const handleDelete = () => remove(id);

  const styles = completed ? { textDecoration: "line-through" } : null;

  // default todo view
  let jsx = (
    <div>
      <li style={styles}>
        {task}
        <button onClick={toggleEdit}>Edit</button>
        <button onClick={toggleCompletion}>âœ“</button>
        <button onClick={handleDelete}>X</button>
      </li>
    </div>
  );

  //todo view when editing
  if (isEditing) {
    jsx = (
      <div>
        <form onSubmit={handleUpdate}>
          <input type="text" value={editTask} onChange={handleChange} />
          <button>Update</button>
        </form>
      </div>
    );
  }

  return jsx;
};

export default Todo;
