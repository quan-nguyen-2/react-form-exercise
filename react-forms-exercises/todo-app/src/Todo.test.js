import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import Todo from "./Todo";

it("renders without crashing", function () {
  render(<Todo />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<Todo />);
  expect(asFragment()).toMatchSnapshot();
});

it("matches snapshot when editing", function () {
  const { asFragment } = render(<Todo />);

  const editBtn = screen.getByText("âœ“");
  fireEvent.click(editBtn);
  expect(asFragment()).toMatchSnapshot();
});

it("calls the update function on form submission", function () {
  const updateMock = jest.fn();
  render(<Todo update={updateMock} />);
  const editBtn = screen.getByText("Edit");
  fireEvent.click(editBtn);
  const updateBtn = screen.getByText("Update");
  fireEvent.click(updateBtn);
  expect(updateMock).toHaveBeenCalled();
});

it("calls the delete function on button click", function () {
  const removeMock = jest.fn();
  render(<Todo remove={removeMock} />);
  const deleteBtn = screen.getByText("X");
  fireEvent.click(deleteBtn);
  expect(removeMock).toHaveBeenCalled();
});
