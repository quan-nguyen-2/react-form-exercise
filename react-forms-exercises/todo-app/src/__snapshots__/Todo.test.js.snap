// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <li>
      <button>
        Edit
      </button>
      <button>
        ✓
      </button>
      <button>
        X
      </button>
    </li>
  </div>
</DocumentFragment>
`;

exports[`matches snapshot when editing 1`] = `
<DocumentFragment>
  <div>
    <li
      style="text-decoration: line-through;"
    >
      <button>
        Edit
      </button>
      <button>
        ✓
      </button>
      <button>
        X
      </button>
    </li>
  </div>
</DocumentFragment>
`;
